### PR TITLE
fix Issue 18627 - Use cephes algorithm for complex divide

### PR DESCRIPTION
**Pull request by ibuclaw**
The current algorithm is based off the Smith algorithm.  This rewrite uses a variant of the Cephes (naive) algorithm that encourages the optimizer to merge the two divides into one (with use of SSE2 `divpd`).

### CODE REVIEW

File: std/complex.d [hunk 1]
```diff
@@ -352,25 +340,12 @@ if (isFloatingPoint!T)
...
-            im = (im*ratio - re)/denom;
-            re = temp;
-        }
-        else
-        {
-            immutable ratio = z.im/z.re;
-            immutable denom = z.re + z.im*ratio;
+        immutable y = re * re + im * im;
+        immutable p = z.re * re + z.im * im;
+        immutable q = z.im * re - z.re * im;
```
Comment (berni44) [inline std/complex.d:345]: IMHO this should be:

```
immutable y = z.re * z.re + z.im * z.im;
immutable p = z.re * re + z.im * im;
immutable q = z.re * im - z.im * re;
```

Comment (ibuclaw) [reply std/complex.d:345]: Yes, I tested the function was working locally, but didn't transpose it correctly into opOpAssign.

**Review by berni44 — APPROVED** [review]

**Comment by ibuclaw** [comment]
Apparently it wasn't quite so straightforward.  But I've been tweaking the algorithm and managed to coerce ldc to produce code that has equivalent performance as native complex division.
https://d.godbolt.org/z/n4P54f

Bad code:
```
    auto prod_re = r * rhs.re;             // cvtsi2sd r; mulsd r, rhs.re;
    auto prod_im = -r * rhs.im;            // neg r; cvtsi2sd r; mulsd r, rhs.re;
    return typeof(return)(prod_re / norm,  // divsd prod_re, norm;
                          prod_im / norm); // divsd prod_im, norm;
```
Good code:
```
    auto prod_re = r * rhs.re;            // cvtsi2sd r;
    auto prod_im = r * -rhs.im;           // mulpd (r, r), (rhs.re, -rhs.im)
    return typeof(return)(prod_re / norm, // divpd (prod_re, prod_im), (norm, norm)
                          prod_im / norm);
```

**Comment by RazvanN7** [comment]
Do you have any measurements to prove the optimization?

**Comment by ibuclaw** [comment]
> Do you have any measurements to prove the optimization?

Yes, and I've been asking other people in slack to verify it as well.

You can run this as a control test (cvtsi2sd+divpd vs 2x cvtsi2sd+divsd)
```
struct Cdouble { double re; double im; }

Cdouble divsd(int a, double b, double c)
{
    asm {
        naked;
        movapd   XMM2, XMM0;
        mulsd    XMM2, XMM0;
        movapd   XMM3, XMM1;
        mulsd    XMM3, XMM1;
        cvtsi2sd XMM4, EDI;
        addsd    XMM3, XMM2;
        neg      EDI;
        xorps    XMM2, XMM2;
        cvtsi2sd XMM2, EDI;
        mulsd    XMM0, XMM4;
        mulsd    XMM1, XMM2;
        divsd    XMM0, XMM3;
        divsd    XMM1, XMM3;
        ret;
    }
}

Cdouble divpd(int a, double b, double c)
{
    asm {
        naked;
        movapd   XMM2, XMM0;
        mulsd    XMM2, XMM0;
        movapd   XMM3, XMM1;
        mulsd    XMM3, XMM1;
        addsd    XMM3, XMM2;
        xorps    XMM2, XMM2;
        cvtsi2sd XMM2, EDI;
        mov      RAX, 0x8000000000000000;
        mov      [RSP-16], RAX;
        movsd    XMM4, [RSP-16];
        xorpd    XMM1, XMM4;
        unpcklpd XMM0, XMM1;
        unpcklpd XMM2, XMM2;
        mulpd    XMM2, XMM0;
        unpcklpd XMM3, XMM3;
        divpd    XMM2, XMM3;
        movapd   XMM1, XMM2;
        unpckhpd XMM1, XMM2;
        movapd   XMM0, XMM2;
        ret;
    }
}

void main()
{
    import std.datetime.stopwatch : benchmark;
    import std.stdio : writefln;

    auto sd = divsd(1, 2.0, 3.0);
    auto pd = divpd(1, 2.0, 3.0);
    assert(sd.re == pd.re && sd.im == pd.im);

    void test_divsd() { divsd(1, 2.0, 3.0); }
    void test_divpd() { divpd(1, 2.0, 3.0); }

    auto result = benchmark!(test_divpd, test_divsd)(100000000);
    writefln("With divpd: %s", result[0]);
    writefln("With divsd: %s", result[1]);
}
```

File: std/complex.d [hunk 2]
```diff
@@ -280,26 +280,42 @@ if (isFloatingPoint!T)
     Complex!(CommonType!(T, R)) opBinaryRight(string op, R)(const R r) const
         if (op == "/" && isNumeric!R)
     {
-        import core.math : fabs;
-        typeof(return) w = void;
-        if (fabs(re) < fabs(im))
+        version (FastMath)
```
Comment (RazvanN7) [inline std/complex.d:283]: Do we want to accommodate the new version as well as the old one? If so, would this require a spec mention?

Comment (ibuclaw) [reply std/complex.d:283]: The fastmath version is less accurate as no range reduction is done.

Comment (RazvanN7) [reply std/complex.d:283]: Ok, but how is the user supposed to choose between the 2? Shouldn't we add a mention in the changelog or in the spec about this?

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**